popde/adaptive_kde.py

```diff
@@ -0,0 +1,191 @@
+    rndgen = np.random.RandomState(seed=1)
+    sample1 = rndgen.normal(mean1, sigma1, size=n_samples)
+    sample2 = rndgen.normal(mean2, sigma2, size=n_samples)
+    sample3 = rndgen.normal(mean3, sigma3, size=n_samples)
+    sample = np.column_stack((sample1, sample2, sample3)) # shape is (n_points, n_features)
+    # Create and fit the adaptive KDE note backend must be set correctly
+    kde = AdaptiveBwKDE(sample, backend='awKDEpy', dim_names=['x', 'y', 'z'], alpha=0.5, input_transf=None)
+    print("kde=", kde)
+    
+    # Generate grid for plotting
+    xgrid = np.linspace(sample1.min(), sample1.max(), 100)
+    ygrid = np.linspace(sample2.min(), sample2.max(), 100)
+    zgrid = np.linspace(sample3.min(), sample3.max(), 100)
+    XX, YY, ZZ = np.meshgrid(xgrid, ygrid, zgrid)
+    eval_pts = np.column_stack((XX.flatten(), YY.flatten(), ZZ.flatten()))
+    kde.fit()
+
+    # Evaluate the KDE at the grid points
+    density_values = kde.evaluate(eval_pts)
+    density_values = density_values.reshape(XX.shape)
+    print(density_values)
+    """
+    def __init__(self, data, backend='KDEpy', bandwidth=1., dim_names=None, alpha=0.5,  input_transf=None, stdize=False, rescale=None):
+        # Additional parameter for the Wang & Wang formula
+        self.alpha = alpha
+        super().__init__(data, input_transf, stdize, rescale,
+                         backend, bandwidth, dim_names)  
+
+    def _calc_local_bandwidth(self, kde_values):
+        """
+        Calculate local bandwidth using the provided function.
+
+        Parameters:
+        -----------
+        kde_values : array-like
+            The KDE values at the data point positions.
+
+        Returns:
+        --------
+        inv_loc_bw : array-like
+            Inverse of the local bandwidth calculated using the provided function.
+        """
+        g = np.exp(np.sum(np.log(kde_values)) / len(kde_values))
+        inv_loc_bw = (kde_values / g) ** self.alpha
+        return inv_loc_bw
+
+    def calculate_per_point_bandwidths(self, pilot_values, adaptive=True):
+        """
+        Calculate per-point bandwidths using the Wang & Wang formula.
+
+        Parameters:
+        -----------
+        pilot_values : array-like
+            The pilot KDE values at the data point positions.
+        adaptive : bool, optional (default=True)
+            If True, use adaptive bandwidth; if False, use fixed bandwidth.
+
+        Returns:
+        --------
+        per_point_bandwidths : array-like
+            Per-point bandwidths calculated using the Wang & Wang formula.
+        """
+        # Calculate the local bandwidth using the provided function
+        local_bandwidths = self._calc_local_bandwidth(pilot_values)
+
+        # Use the local bandwidths to calculate per-point bandwidths
+        if adaptive:
+            per_point_bandwidths = self.bandwidth * local_bandwidths
+        else:
+            per_point_bandwidths = np.ones_like(local_bandwidths) / self.bandwidth
```

Comment:
here, you are _dividing_ by self.bandwidth.  This doesn't seem consistent.  If you want `per_point_bandwidths` to behave like the global bw variable, it should be * not /

Reply:
removed else